October 2020 Release Notes (#39)
livingdocsIO/livingdocs-release-notes: releases/release-2020-10.md

`@livingdocs/editor` | `release-2020-10`

## Livingdocs Server
How to require the server in your package.json:
```json
"dependencies": {
  "@livingdocs/server": "release-2020-10",
}
```

- Link to the release branch:
  https://github.com/livingdocsIO/livingdocs-server/tree/release-2020-10

### Livingdocs Server Patches
- [v??.?.?](https://github.com/livingdocsIO/livingdocs-server/releases/tag/v??.?.?): text

## Livingdocs Editor
How to require the editor in your package.json:
```json
"dependencies": {
  "@livingdocs/editor": "release-2020-10",
}
```

- Link to the release branch:
  https://github.com/livingdocsIO/livingdocs-editor/tree/release-2020-10

### Livingdocs Editor Patches
- [v??.?.?](https://github.com/livingdocsIO/livingdocs-editor/releases/tag/v??.?.?): text

# Highlights

## New APIs :tada:

We added numerous APIs for the 'public API', 'livingdocs-server CLI' and we have good progress with the Angular to Vue
transition. Find more in the [APIs](#apis-gift) section of the release notes.

## WoodWing Assets Integration :tada:

Adds a basic WoodWing integration into Livingdocs:
- Support drag+drop from WoodWing assets to Livingdocs

> - Proof of concept: Whenever an image is uploaded to Livingdocs, it's also uploaded to WoodWing assets with some basic metadata.

Reviewer: why do you call it proof of concept?

Author: Because you had this passage in your PR :-)

References:
  * [Video](https://vimeo.com/444823016)
  * [Editor PR](https://github.com/livingdocsIO/livingdocs-editor/pull/3771)
  * [Server PR](https://github.com/livingdocsIO/livingdocs-server/pull/3079)

> 

Reviewer: I would mention the Livingdocs to WoodWing Studio project as a sidenote here:
```
Related: we also implemented an open-source boilerplate for a Livingdocs to WoodWing Studio transformer as an AWS serverless app. Find the code on our Github repo: https://github.com/livingdocsIO/livingdocs-to-woodwing-exporter
```

## Referenced Documents :tada:

We harmonized the visual appearance for document references in the editor e.g.
- media library document references
- 'copy of' / 'embedded in' on the publish screen
- ...more will follow in the next releases

References:
  * [Editor PR](https://github.com/livingdocsIO/livingdocs-editor/pull/3796)
  * [Server PR](https://github.com/livingdocsIO/livingdocs-server/pull/3140)

## Parallel Image Upload :tada:

We massively improved the image uploading process. The improvements are:
- Upload multiple images in parallel
- Edit metadata for multiple images at once
- Cancel single uploads

Look at this [PR](https://github.com/livingdocsIO/livingdocs-editor/pull/3814) for some impressions.

> 

Reviewer: I remember seeing an animated gif of this somewhere. Would be nice to add this here

Author: @benib is there somewhere an animated gif?

## Image Service 2.0 :tada:

[Image services 2.0](https://docs.livingdocs.io/evaluation-guide/image-services) have experienced a lot of improvements.

> For example register custom image services in in the downstream (instead of adding them to the `livingdocs-framework`).

Reviewer: That is the worst description I've ever read 😆 
- I don't know what "Image Service 2.0" is in the first place -> that should go into the description
- Never start a description with "for example"

Other nice improvements are:

> - The rendering is separated from the imageService itself and renderStrategies can be defined via configuration.

Reviewer: an example code block would help a lot

Author: We don't want to blow up the release notes. Examples for new features should be in our documentation. I asked already @peyerluk to add some documentation.

> - New image service configuration like a `imgTagRenderStrategy` are available

Reviewer: Please list all of them and what they do

Author: Same here, not in the release notes.

> - Added a backgroundImage render strategy with responsive behaviour

Reviewer: I know for a fact that responsive bg images are hard. Please describe which strategy you used here

Author: //cc @peyerluk

> Look [here](https://github.com/livingdocsIO/livingdocs-framework/pull/497) fore more infos on how the new image services can be configured.

Reviewer: We should def. do documentation for this and not just have a PR

Author: It's already ordered.

Pull Request: [livingdocs-server #3086](https://github.com/livingdocsIO/livingdocs-server/pull/3086)

- ⏳ Deprecated static server config `assetManagement`. Rename `assetManagement` to `mediaLibrary`
- ⏳ Remove property `assetManagement.enabled` (currently it only throws a deprecation message)
- ⏳ Remove property `assetManagement.googleVision` (currently it only throws a deprecation message)

## Deprecate MediaLibrary paginationSize

Pull Request: [livingdocs-server #3092](https://github.com/livingdocsIO/livingdocs-server/pull/3092)

- ⏳ deprecate static server config `mediaLibrary.paginationSize`. The config will be removed in the next releases but has already no effect anymore. We use an internal default of 35.

## Data Migrations

Pull Request: [livingdocs-server #3151](https://github.com/livingdocsIO/livingdocs-server/pull/3151) :gift:

- ⏳ `grunt data-migration-create-and-prepare` should not be used anymore. The task has been replaced with `livingdocs-server data-migration-run`
- ⏳ `grunt data-migration-accept` should not be used anymore. The task has been replaced with `livingdocs-server data-migration-run`

# APIs :gift:

## Register Custom Vue Display Filter :gift:

> One can register and use a custom display filter as a Vue component (e.g. on a dashboard).

Reviewer 2: ```suggestion
One can register a Vue component as a custom display filter (e.g. on a dashboard).
```

References:
  * [Documentation](https://github.com/livingdocsIO/livingdocs/pull/317)
  * [Editor PR](https://github.com/livingdocsIO/livingdocs-editor/pull/3777)

## Doc Includes with Vue :gift:

> Implement [Doc Includes](https://docs.livingdocs.io/evaluation-guide/intro#summary-of-a-doc-include) with Vue.

Reviewer 2: Since the thing one can do with Vue is the editor customization part, should we link to https://docs.livingdocs.io/reference-docs/includes/editor_customization ? instead
And: we need to update the docs before these release notes are sent out.

Author: I would love it!

Author: Have you time for that in the next few days @benib ?

References:
  * [Documentation (Twitter Example)](https://github.com/livingdocsIO/livingdocs/pull/312)
  * [Editor PR](https://github.com/livingdocsIO/livingdocs-editor/pull/3768)
  * [Server PR](https://github.com/livingdocsIO/livingdocs-server/pull/3098)

## Public API - Media Library Endpoints :gift:

The public API contains new endpoints for the media library.

Examples:
- `GET /api/v1/mediaLibrary/:id`
- `GET /api/v1/mediaLibrary?ids=1,2,3` or `GET /api/v1/mediaLibrary?externalId=foo&systemName=comyan`

References:
  * Documentation - 'https://your-editor.com/public-api'
  * [Editor PR](https://github.com/livingdocsIO/livingdocs-editor/pull/3099)

## Public API - Expose Public API via the server API :gift:

Get access to the public API via server API
- `const publicApi = liServer.features.api('li-public-api')`

References:
  * [Server PR](https://github.com/livingdocsIO/livingdocs-server/pull/3138)

> ## Public API - Support client defined documentId in the import API #3145 :gift:

Reviewer: This needs documentation (update) as well

Author: @marcbachmann can you write some documentation (update) for that?

## Livingdocs-server CLI - Improve reindex tasks :gift:

New options are supported
- `livingdocs-server es-search-reindex` supports a project filter, e.g. `--project=magazine`
- `livingdocs-server es-publication-reindex` supports a project filter, e.g. `--project=magazine`
- `livingdocs-server es-publication-reindex` supports a contentType filter, e.g. `--content-type=regular`

References:
  * [Server PR](https://github.com/livingdocsIO/livingdocs-server/pull/3162)

# Internal Changes

Changes that should not affect customers.
But some customers use internal functions and can be affected by this changes.

## CSS Changes

```
// renamed CSS classes
ld-filter-toolbar -> li-filter-toolbar
ld-filter.is-disabled renamed to li-filter-group__item--disabled
ld-filter.is-set renamed to li-filter-group__item--active

// removed CSS classes
ld-filter
ld-filter-options
```

PR: [livingdocs-editor #3801](https://github.com/livingdocsIO/livingdocs-editor/pull/3801)

# Other Changes

### Design

* A lot of small UI hint additions [livingdocs-editor #3835](https://github.com/livingdocsIO/livingdocs-editor/pull/3835) :gift:

### Improvements

* Media Library

>   * Show message when no document has an image reference [livingdocs-editor #3826](https://github.com/livingdocsIO/livingdocs-editor/pull/3826) :gift:

Reviewer 2: ```suggestion
  * Show message when an image is not used in any document [livingdocs-editor #3826](https://github.com/livingdocsIO/livingdocs-editor/pull/3826) :gift:
```

  * Show proper message when image is broken [livingdocs-editor #3828](https://github.com/livingdocsIO/livingdocs-editor/pull/3828) :gift:
* Imatrics

>   * Store any concept properties returned from the server [livingdocs-editor #3837](https://github.com/livingdocsIO/livingdocs-editor/pull/3837) :gift:

Reviewer 2: ```suggestion
  * iMatrics: Store any concept properties returned from the server [livingdocs-editor #3837](https://github.com/livingdocsIO/livingdocs-editor/pull/3837) :gift:
```

Author: Its a sub menu of Imatrics, therefore it makes no sense to add an Imatrics prefix again.

>   * Updates suggestion in realtime [livingdocs-editor #3817](https://github.com/livingdocsIO/livingdocs-editor/pull/3817) :gift:

Reviewer 2: ```suggestion
  * iMatrics: Update suggestion in realtime in the dashboard [livingdocs-editor #3817](https://github.com/livingdocsIO/livingdocs-editor/pull/3817) :gift:
```

## WoodWing Assets Integration :tada:

Adds a basic WoodWing integration into Livingdocs:
- Support drag+drop from WoodWing assets to Livingdocs
- Proof of concept: Whenever an image is uploaded to Livingdocs, it's also uploaded to WoodWing assets with some basic metadata.

References:
  * [Video](https://vimeo.com/444823016)
  * [Editor PR](https://github.com/livingdocsIO/livingdocs-editor/pull/3771)
  * [Server PR](https://github.com/livingdocsIO/livingdocs-server/pull/3079)

Related: We also implemented an open-source boilerplate for a [Livingdocs to WoodWing Studio transformer](https://github.com/livingdocsIO/livingdocs-to-woodwing-exporter) as an AWS serverless app.

## Referenced Documents :tada:

We harmonized the visual appearance for document references in the editor e.g.
- media library document references
- 'copy of' / 'embedded in' on the publish screen
- ...more will follow in the next releases

References:
  * [Editor PR](https://github.com/livingdocsIO/livingdocs-editor/pull/3796)
  * [Server PR](https://github.com/livingdocsIO/livingdocs-server/pull/3140)

## Parallel Image Upload :tada:

> We massively improved the image uploading process. The improvements are:

Reviewer 3: -> We improved the image uploading process for cases when the user needs to provide image metadata before an image can be uploaded.
